gtsam/basis/Basis.h

```diff
@@ -0,0 +1,373 @@
+   * This functor is used to evaluate a parameterized function at a given scalar
+   * value x. When given a specific M*N parameters, returns an M-vector the M
+   * corresponding functions at x, possibly with Jacobian wrpt the parameters.
+   */
+  /// Manifold interpolation
+  template <class T>
+  class ManifoldEvaluationFunctor
+      : public VectorEvaluationFunctor<traits<T>::dimension> {
+    enum { M = traits<T>::dimension };
+    using Base = VectorEvaluationFunctor<M>;
+
+   public:
+    // Used by FunctorizedFactor
+    using argument_type = Matrix;
+    using return_type = T;
+
+    /// Default Constructor
+    ManifoldEvaluationFunctor(size_t N, double x) : Base(N, x) {}
+
+    /// Constructor, with interval [a,b]
+    ManifoldEvaluationFunctor(size_t N, double x, double a, double b)
+        : Base(N, x, a, b) {}
+
+    /// Manifold evaluation
+    T apply(const typename Base::MatrixMN& F,
+            OptionalJacobian</*MxMN*/ -1, -1> H = boost::none) const {
+      // Interpolate the M-dimensional vector to yield a vector in tangent space
+      Eigen::Matrix<double, M, 1> xi = Base::operator()(F, H);
+
+      // Now call retract with this M-vector, possibly with derivatives
+      Eigen::Matrix<double, M, M> D_result_xi;
+      T result = T::ChartAtOrigin::Retract(xi, H ? &D_result_xi : 0);
+
+      // Finally, if derivatives are asked, apply chain rule where H is Mx(M*N)
+      // derivative of interpolation and D_result_xi is MxM derivative of
+      // retract.
+      if (H) *H = D_result_xi * (*H);
+
+      // and return a T
+      return result;
+    }
+
+    /// c++ sugar
+    T operator()(const typename Base::MatrixMN& F,
+                 OptionalJacobian</*M*N*/ -1, -1> H = boost::none) const {
+      return apply(F, H);  // might call apply in derived
+    }
+  };
+
+  /// Base class for functors below that calculates weights
+  class DerivativeFunctorBase {
+   protected:
+    typename DERIVED::Weights weights_;
+
+   public:
+    DerivativeFunctorBase(size_t N, double x)
+        : weights_(DERIVED::DerivativeWeights(N, x)) {}
+
+    DerivativeFunctorBase(size_t N, double x, double a, double b)
+        : weights_(DERIVED::DerivativeWeights(N, x, a, b)) {}
+
+    void print(const std::string& s = "") const {
+      std::cout << s << (s != "" ? " " : "") << weights_ << std::endl;
+    }
+  };
+
+  /// Given values f at the Chebyshev points, predict derivative at x
+  class DerivativeFunctor : protected DerivativeFunctorBase {
+   public:
+    // Used by FunctorizedFactor
+    using argument_type = typename DERIVED::Parameters;
+    using return_type = double;
+
+    DerivativeFunctor(size_t N, double x) : DerivativeFunctorBase(N, x) {}
+
+    DerivativeFunctor(size_t N, double x, double a, double b)
+        : DerivativeFunctorBase(N, x, a, b) {}
+
+    double apply(const typename DERIVED::Parameters& f,
+                 OptionalJacobian</*1xN*/ -1, -1> H = boost::none) const {
```

Comment:
Since 1 is known at compile time, perhaps do: `OptionalJacobian< 1, -1>` ?

Reply:
Kept it like this for consistency. This won't be a major improvement in efficiency since we barely use this particular functor.

Reply:
Updated to `OptionalJacobian<1, -1>`.

Reply:
I realized why I didn't do this before (and I am certain I gave Frank the same reasoning): `OptionalJacobian` doesn't support mixed static and dynamic sizes. We can either have it be fully static aka `OptionalJacobian<1, N>` or completely dynamic (which is what I have done).

Adding this support is not on my A->B and there is no real need for it per se since the ROI isn't valuable enough IMO.